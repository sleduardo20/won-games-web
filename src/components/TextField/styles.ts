import styled, { css } from 'styled-components';
import { TextFieldProps } from '.';

export const Container = styled.div``;

export const Input = styled.input`
  ${({ theme }) => css`
    color: ${theme.colors.black};
    font-family: ${theme.font.family};
    outline: none;
    border: 0;
    padding: ${theme.spacings.xxsmall} 0;
    background: transparent;
    flex: 1;
  `}
`;

export const Label = styled.label`
  ${({ theme }) => css`
    font-size: ${theme.font.sizes.small};
    color: ${theme.colors.black};
    cursor: pointer;
  `}
`;

export const Icon = styled.div`
  display: flex;
  width: 2rem;
  color: ${({ theme }) => theme.colors.gray};

  & > svg {
    width: 100%;
  }
`;

const wrapperModifiers = {
  left: () => css`
    ${Icon} {
      margin-right: 0.4rem;
    }
  `,

  rigth: () => css`
    flex-direction: row-reverse;
    ${Icon} {
      margin-left: 0.4rem;
    }
  `,
};

export const Wrapper = styled.div<Pick<TextFieldProps, 'iconPosition'>>`
  ${({ theme, iconPosition }) => css`
    display: flex;
    background: ${theme.colors.ligthGray};
    border-radius: 0.2rem;
    padding: 0 ${theme.spacings.xsmall};
    border: 0.2rem solid ${theme.colors.ligthGray};

    ${iconPosition && wrapperModifiers[iconPosition]};

    &:focus-within {
      box-shadow: 0 0 0.5rem ${theme.colors.primary};
    }
  `}
`;
