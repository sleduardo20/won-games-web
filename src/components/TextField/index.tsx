import { InputHTMLAttributes, useState } from 'react';
import { Container, Wrapper, Input, Label, Icon, Error } from './styles';

export type TextFieldProps = {
  onInput?: (value: string) => void;
  label?: string;
  initialValue?: string;
  icon?: React.ReactNode;
  disabled?: boolean;
  iconPosition?: 'left' | 'rigth';
  error?: string;
} & InputHTMLAttributes<HTMLInputElement>;

const TextField = ({
  onInput,
  label,
  name,
  initialValue = '',
  icon,
  disabled = false,
  iconPosition = 'left',
  error,
  ...rest
}: TextFieldProps) => {
  const [value, setValue] = useState(initialValue);

  const handleOnChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const newValue = event.currentTarget.value;
    setValue(newValue);

    if (onInput) {
      onInput(newValue);
    }
  };

  return (
    <Container iconPosition={iconPosition} disabled={disabled} error={!!error}>
      {!!label && <Label htmlFor={name}>{label}</Label>}
      <Wrapper>
        {!!icon && <Icon>{icon}</Icon>}
        <Input
          type="text"
          onChange={handleOnChange}
          value={value}
          disabled={disabled}
          name={name}
          {...(label ? { id: name } : {})}
          {...rest}
        />
      </Wrapper>
      {!!error && <Error>{error}</Error>}
    </Container>
  );
};

export default TextField;
