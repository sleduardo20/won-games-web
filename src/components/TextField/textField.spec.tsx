import { screen, waitFor } from '@testing-library/react';
import { Mail } from 'styled-icons/material-outlined';
import userEvent from '@testing-library/user-event';
import { renderWithTheme } from 'utils/tests/helpers';

import TextField from '.';

describe('<TextField/>', () => {
  it('should be able render TextField component with label', () => {
    renderWithTheme(<TextField id="field" labelFor="field" label="label" />);

    expect(screen.getByLabelText('label')).toBeInTheDocument();
  });

  it('should be able render TextField component without label', () => {
    renderWithTheme(<TextField />);

    expect(screen.queryByText('label')).not.toBeInTheDocument();
  });

  it('should be able render TextField component with placeholder', () => {
    renderWithTheme(<TextField placeholder="email" />);

    expect(screen.getByPlaceholderText('email')).toBeInTheDocument();
  });

  it('should be able changes its value when typing', async () => {
    const onInput = jest.fn();

    renderWithTheme(
      <TextField id="field" labelFor="field" label="label" onInput={onInput} />,
    );

    const input = screen.getByRole('textbox');
    const text = 'new text';

    userEvent.type(input, text);

    await waitFor(() => {
      expect(input).toHaveValue(text);
      expect(onInput).toHaveBeenCalledTimes(text.length);
    });

    expect(onInput).toBeCalledWith(text);
  });

  it('should be able accessible by tab', () => {
    renderWithTheme(
      <TextField id="textfield" label="textfield" labelFor="textfield" />,
    );

    const input = screen.getByLabelText('textfield');

    expect(document.body).toHaveFocus();

    userEvent.tab();

    expect(input).toHaveFocus();
  });

  it('should be able render TextField component with icon', () => {
    renderWithTheme(
      <TextField placeholder="email" icon={<Mail data-testid="icon" />} />,
    );

    expect(screen.getByTestId('icon')).toBeInTheDocument();
  });

  it('should be able render TextField component with icon position left', () => {
    renderWithTheme(
      <TextField icon={<Mail data-testid="icon" />} iconPosition="left" />,
    );

    const wrapper = screen.getByRole('textbox').parentElement;

    expect(wrapper).toHaveStyle({ display: 'flex' });
  });

  it('should be able render TextField component with icon position rigth', () => {
    renderWithTheme(
      <TextField icon={<Mail data-testid="icon" />} iconPosition="rigth" />,
    );

    const wrapper = screen.getByRole('textbox').parentElement;

    expect(wrapper).toHaveStyle({ flex: 'row-reverse' });
  });
});
