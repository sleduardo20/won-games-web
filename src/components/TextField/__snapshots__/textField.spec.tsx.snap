// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField/> should be able render TextField component with error 1`] = `
.c7 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c8 {
  color: #030517;
  font-family: Poppins,--apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  outline: none;
  border: 0;
  padding: 0.8rem 0;
  background: transparent;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c8:-webkit-autofill {
  -webkit-box-shadow: 0 0 0 2.4rem #EAEAEA inside;
  -webkit-filter: none;
  filter: none;
}

.c2 {
  font-size: 1.4rem;
  color: #030517;
  cursor: pointer;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 2rem;
  color: #8f8f8f;
}

.c6 > svg {
  width: 100%;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background: #EAEAEA;
  border-radius: 0.2rem;
  padding: 0 1.6rem;
  border: 0.2rem solid #EAEAEA;
}

.c4:focus-within {
  box-shadow: 0 0 0.5rem #f231a5;
}

.c9 {
  color: #FF6347;
  font-size: 1.2rem;
}

.c0 .c5 {
  margin-right: 0.4rem;
}

.c0 .c3 {
  border-color: #FF6347;
}

.c0 .c5,
.c0 .c1 {
  color: #FF6347;
}

<div>
  <div
    class="c0"
  >
    <label
      class="c1 c2"
      for="textfield"
    >
      textfield
    </label>
    <div
      class="c3 c4"
    >
      <div
        class="c5 c6"
      >
        <svg
          aria-hidden="true"
          class="c7"
          data-testid="icon"
          fill="currentColor"
          focusable="false"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0V0z"
            fill="none"
          />
          <path
            d="M22 6c0-1.1-.9-2-2-2H4c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6zm-2 0l-8 4.99L4 6h16zm0 12H4V8l8 5 8-5v10z"
          />
        </svg>
      </div>
      <input
        class="c8"
        id="textfield"
        name="textfield"
        type="text"
        value=""
      />
    </div>
    <span
      class="c9"
    >
      Error Message
    </span>
  </div>
</div>
`;
